### PR TITLE
Feature: Add dummy js client

### PR DESCRIPTION
**Pull request by josemarinas**
## Description

Add dummy JS client package

[OS-465](https://aragonassociation.atlassian.net/browse/OS-465)

[OS-465]: https://aragonassociation.atlassian.net/browse/OS-465?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/js-client/.gitignore [hunk 1]
```diff
@@ -0,0 +1,5 @@
+*.log
+.DS_Store
+node_modules
+dist
+*.tgz
```
Comment (heueristik) [inline packages/js-client/.gitignore]:
```suggestion
*.tgz

```

File: packages/js-client/package.json [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    "access": "public"
+  },
+  "engines": {
+    "node": ">=16"
+  },
+  "scripts": {
+    "start": "tsdx watch",
+    "build": "tsdx build",
+    "test": "TESTING=true tsdx test --verbose --runInBand",
+    "lint": "tsdx lint",
```
Comment (heueristik) [inline packages/js-client/package.json]: Seems to not work at least on the branch I checked out.

File: packages/js-client/package.json [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+  "publishConfig": {
+    "access": "public"
+  },
+  "engines": {
+    "node": ">=16"
+  },
+  "scripts": {
+    "start": "tsdx watch",
+    "build": "tsdx build",
+    "test": "TESTING=true tsdx test --verbose --runInBand",
```
Comment (heueristik) [inline packages/js-client/package.json:21]: I get some warnings

```
 PASS  test/integration/methods.test.ts
  Methods
    ✓ should test a method (3ms)

Test Suites: 5 passed, 5 total
Tests:       14 passed, 14 total
Snapshots:   0 total
Time:        9.731s
Ran all test suites.
  console.warn
    Method not implemented.

       8 |   // implementation of the methods in the interface
       9 |   public myMethod() {
    > 10 |     console.warn('Method not implemented.');
         |             ^
      11 |   }
      12 | }
      13 |

      at SimpleStorageClientMethods.Object.<anonymous>.SimpleStorageClientMethods.myMethod (src/internal/modules/methods.ts:10:13)
      at test/integration/methods.test.ts:19:20
      at step (node_modules/tslib/tslib.js:193:27)
      at Object.next (node_modules/tslib/tslib.js:174:57)
      at node_modules/tslib/tslib.js:167:75
      at Object.__awaiter (node_modules/tslib/tslib.js:163:16)

✨  Done in 10.81s.
```

File: packages/js-client/package.json [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    "lint": "tsdx lint",
+    "prepare": "tsdx build",
```
Comment (heueristik) [inline packages/js-client/package.json]:
```suggestion
    "prepare": "yarn build",
```
Why the duplication when the two are identical?

File: packages/js-client/package.json [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    "size": "size-limit",
```
Comment (heueristik) [inline packages/js-client/package.json:24]:
```
➜  js-client git:(f/OS-465-dummy-js-client) yarn size
yarn run v1.22.19
$ size-limit
✔ Adding to empty esbuild project
  
  
  Size Limit can’t find files at dist/plugin-client.cjs.production.min.js
  
  Size Limit can’t find files at dist/plugin-client.esm.js
error Command failed with exit code 1.
```

File: packages/js-client/package.json [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    "analyze": "size-limit --why",
```
Comment (heueristik) [inline packages/js-client/package.json:25]:
```
➜  js-client git:(f/OS-465-dummy-js-client) yarn analyze
yarn run v1.22.19
$ size-limit --why
 ERROR  Argument --why works only with @size-limit/webpack plugin and @size-limit/webpack-why plugin.
        You can add Bundle Analyzer to you own bundler.
error Command failed with exit code 1.
```

File: packages/js-client/package.json (hunk 3 above)
Comment (josemarinas) [reply packages/js-client/package.json:21]: Yes, I left some warnings when calling the methods, since it's just a template I wanted to have some methods examples butt without doing anything

File: packages/js-client/package.json (hunk 4 above)
Comment (josemarinas) [reply packages/js-client/package.json]: This is generated by tsdx. Its true that is exactly the samee thing so i'll delete it

File: .github/workflows/js-client-update-docs.yml [hunk 7]
```diff
@@ -0,0 +1,65 @@
+name: Documentation Update
```
Comment (brickpop) [inline .github/workflows/js-client-update-docs.yml]: Will we need this file for the template?

File: .github/workflows/js-client-update-docs.yml [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+      - 'packages/js-client/**'
+      - '.github/workflows/js-client-*.yml'
+
+jobs:
+  update_docs:
+    if: github.actor != 'arabot-1'
+    permissions: read-all
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout sdk repo
```
Comment (brickpop) [inline .github/workflows/js-client-update-docs.yml]: Relevant?

File: .github/workflows/js-client-update-docs.yml [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+      - name: Checkout developer-portal repo
+        uses: actions/checkout@v3
+        with:
+          repository: aragon/developer-portal
+          ref: staging
+          path: developer-portal
+          # token: ${{ secrets.ARABOT_PAT }}
+      - name: Clean the docs/sdk folder in the developer-portal
+        run: rm -rf $GITHUB_WORKSPACE/developer-portal/docs/sdk
+      - name: Copy the docs folder to the developer-portal
```
Comment (brickpop) [inline .github/workflows/js-client-update-docs.yml]: Relevant?

File: packages/js-client/README.md [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+
+#### Setup Files
+
+This is the folder structure we set up for you:
+
+```txt
+/src
+  index.tsx       # EDIT THIS
+/test
+  blah.test.tsx   # EDIT THIS
```
Comment (brickpop) [inline packages/js-client/README.md:45]: Not edited by us, but not applicable here

File: packages/js-client/package.json [hunk 11]
```diff
@@ -0,0 +1,75 @@
+{
+  "name": "@aragon/sdk-client-template",
```
Comment (brickpop) [inline packages/js-client/package.json]:
```suggestion
  "name": "@aragon/plugin-js-client-template",
```

File: packages/js-client/README.md [hunk 12]
```diff
@@ -0,0 +1,103 @@
+# TSDX User Guide
```
Comment (brickpop) [inline packages/js-client/README.md:1]: Shouldn't the readme explain how to use the existing client and how to adapt things? Or is it intended for later?

File: packages/js-client/src/context.ts [hunk 13]
```diff
@@ -0,0 +1,71 @@
+import { SimpleStorageContextParams } from './types';
+import { Context, ContextCore, ContextState } from '@aragon/sdk-client-common';
+
+// TODO
+// import from @aragon/sdk-client-common
```
Comment (brickpop) [inline packages/js-client/src/context.ts]: ?

File: packages/js-client/src/context.ts [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+
+type SimpleStorageContextState = ContextState & {
+  // extend the Context state with a new state for storing
+  // the new parameters
+  myParam: string;
+};
+
+type SimpleStorageOverridenState = OverriddenState & {
+  myParam: boolean;
+};
```
Comment (brickpop) [inline packages/js-client/src/context.ts]: Why is this needed? The top section seems to have more boilerplate than necessary

File: packages/js-client/src/context.ts [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+    // so we don't need to call it again
+    super(contextParams);
+    if (context) {
+      // copy the context properties to this
+      Object.assign(this, context);
+    }
+    if (contextParams) {
+      // overide the context params with the ones passed to the constructor
+      this.set(contextParams ?? {});
+    }
```
Comment (brickpop) [inline packages/js-client/src/context.ts]: That might be confusing for a newcomer. I would make the difference clear with something like:
1. `aragonContext`
2. `pluginContext`

File: packages/js-client/src/context.ts [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+    // set the default values for the new params
+    this.setDefaults();
+    // override default params if specified in the context
+    if (contextParams.myParam) {
+      // override the myParam value
+      this.state.myParam = contextParams.myParam;
+      // set the overriden flag to true in case set is called again
+      this.overriden.myParam = true;
+    }
+  }
```
Comment (brickpop) [inline packages/js-client/src/context.ts]: Shoudn't this be consistent with the constructor? or at least detect which field is handled by whom?

**Review by brickpop — COMMENTED** [review]
First round

File: packages/js-client/src/internal/interfaces.ts [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+  // fill with methods for estimation
+  myMethod(): void;
+}
+export interface ISimpleStorageClientEncoding {
+  // fill with methods for encoding actions
+  myAction(): void;
+}
+export interface ISimpleStorageClientDecoding {
+  // fill with methods for decoding actions
+  myAction(): void;
```
Comment (brickpop) [inline packages/js-client/src/internal/interfaces.ts]: why not using real method names to use the plugin?
- `getValue()`
- `setValueAction(val)`

File: packages/js-client/src/internal/modules/decoding.ts [hunk 18]
```diff
@@ -0,0 +1,12 @@
...
+import { ClientCore } from '@aragon/sdk-client-common';
+
+export class SimpleStorageClientDecoding
+  extends ClientCore
+  implements ISimpleStorageClientDecoding
+{
+  // implementation of the methods in the interface
+  public myAction() {
+    console.warn('Method not implemented.');
+  }
```
Comment (brickpop) [inline packages/js-client/src/internal/modules/decoding.ts]: As said before, I would provide functional code doing real things. With a big comment telling people where to implement their own logic

File: packages/js-client/src/internal/modules/encoding.ts [hunk 19]
```diff
@@ -0,0 +1,12 @@
+import { ISimpleStorageClientEncoding } from '../interfaces';
+import { ClientCore } from '@aragon/sdk-client-common';
+
+export class SimpleStorageClientEncoding
+  extends ClientCore
+  implements ISimpleStorageClientEncoding
+{
+  // implementation of the methods in the interface
+  public myAction() {
```
Comment (brickpop) [inline packages/js-client/src/internal/modules/encoding.ts]: Same

File: packages/js-client/src/internal/modules/estimation.ts [hunk 20]
```diff
@@ -0,0 +1,12 @@
+import { ISimpleStorageClientEstimation } from '../interfaces';
+import { ClientCore } from '@aragon/sdk-client-common';
+
+export class SimpleStoragClientEstimation
+  extends ClientCore
+  implements ISimpleStorageClientEstimation
+{
+  // implementation of the methods in the interface
+  public myMethod() {
```
Comment (brickpop) [inline packages/js-client/src/internal/modules/estimation.ts]: Same

File: packages/js-client/src/internal/modules/methods.ts [hunk 21]
```diff
@@ -0,0 +1,12 @@
+import { ISimpleStorageClientMethods } from '../interfaces';
+import { ClientCore } from '@aragon/sdk-client-common';
+
+export class SimpleStorageClientMethods
+  extends ClientCore
+  implements ISimpleStorageClientMethods
+{
+  // implementation of the methods in the interface
+  public myMethod() {
```
Comment (brickpop) [inline packages/js-client/src/internal/modules/methods.ts]: Same

File: packages/js-client/src/types.ts [hunk 22]
```diff
@@ -0,0 +1,7 @@
+import { ContextParams } from '@aragon/sdk-client-common';
+
+export type SimpleStorageContextParams = ContextParams & {
+  // optional so we can set default values for the parameter
+  myParam?: string;
+  // add custom params
+};
```
Comment (brickpop) [inline packages/js-client/src/types.ts]: Isn't this redundant with the `context.ts` type definitions?

File: packages/js-client/test-environment.js [hunk 23]
```diff
@@ -0,0 +1,16 @@
+const Environment = require('jest-environment-jsdom');
+
+module.exports = class CustomTestEnvironment extends Environment {
```
Comment (brickpop) [inline packages/js-client/test-environment.js:3]: Why is this needed? to run the IPFS client on NodeJS as if it was a browser?

File: packages/js-client/test/mocks/aragon-sdk-ipfs.ts [hunk 24]
```diff
@@ -0,0 +1,23 @@
+export const mockedIPFSClient = {
```
Comment (brickpop) [inline packages/js-client/test/mocks/aragon-sdk-ipfs.ts:11]: I would add some comment here, explaining to devs that these are default values and also how to add their own desired ones ad-hoc.

File: packages/js-client/test/mocks/graphql-request.ts [hunk 25]
```diff
@@ -0,0 +1,26 @@
+import { gql, GraphQLClient } from 'graphql-request';
```
Comment (brickpop) [inline packages/js-client/test/mocks/graphql-request.ts:13]: Same as above, brief intro + how to mock in-place

File: packages/js-client/test/unit/context.test.ts [hunk 26]
```diff
@@ -0,0 +1,291 @@
...
+    );
+    context.graphql.map((graphqlClient) =>
+      expect(graphqlClient).toBeInstanceOf(GraphQLClient)
+    );
+    expect(context.myParam).toBe('notDefault');
+  });
+  it('Should set a new context and have the correct values', () => {
+    const context = new SimpleStorageContext(contextParams);
+    contextParams = {
+      network: 'goerli',
```
Comment (brickpop) [inline packages/js-client/test/unit/context.test.ts:90]: SupportedNetworks.GOERLI ?

**Review by brickpop — COMMENTED** [review]
Round 2

File: .github/workflows/js-client-update-docs.yml (hunk 7 above)
Comment (josemarinas) [reply .github/workflows/js-client-update-docs.yml]: Not sure, I copied all the workflows from the sddk and aadapted the routes where neceessary. I'm not sure if it will be on the template because it is not absolutelly necesary for an external developer, but all our plugins are going to need it

File: .github/workflows/js-client-update-docs.yml (hunk 8 above)
Comment (josemarinas) [reply .github/workflows/js-client-update-docs.yml]: same

File: packages/js-client/README.md (hunk 10 above)
Comment (josemarinas) [reply packages/js-client/README.md:45]: Yep I diddnt wrote the readme yet, I wanted to focus on finishing the code paarte before explaining how to work with it

File: packages/js-client/package.json (hunk 11 above)
Comment (josemarinas) [reply packages/js-client/package.json]: I would change it to `@aragon/js-client-template` because its the name of tthe folder @brickpop

File: packages/js-client/README.md (hunk 12 above)
Comment (josemarinas) [reply packages/js-client/README.md:1]: Same as before, first I wsaant to end the code then write the readme

File: packages/js-client/src/context.ts (hunk 13 above)
Comment (josemarinas) [reply packages/js-client/src/context.ts]: I forgot to add this to the client-common package, so this is intended to be imported from there

File: packages/js-client/src/context.ts (hunk 14 above)
Comment (josemarinas) [reply packages/js-client/src/context.ts]: This is just an examplo of how the extend the `OverridenState`, the `OverridenState` type itself should be imported from the `client-common` package

File: packages/js-client/src/context.ts (hunk 15 above)
Comment (josemarinas) [reply packages/js-client/src/context.ts]: not sure what do you mean, rename context to `aragonContext` ?

File: packages/js-client/src/context.ts (hunk 16 above)
Comment (josemarinas) [reply packages/js-client/src/context.ts]: The `super` in the contructor calls this function instead of the parent one because its overriden, thats why the super.set, all the new params are handled by this set

File: packages/js-client/src/internal/interfaces.ts (hunk 17 above)
Comment (josemarinas) [reply packages/js-client/src/internal/interfaces.ts]: 👍

File: packages/js-client/src/internal/modules/decoding.ts (hunk 18 above)
Comment (josemarinas) [reply packages/js-client/src/internal/modules/decoding.ts]: Mmm okay, I caan implement the functions for the Simple storage plugin

File: packages/js-client/src/types.ts (hunk 22 above)
Comment (josemarinas) [reply packages/js-client/src/types.ts]: Mmm nope, this is for adding a new property to the context params if we want

File: packages/js-client/test-environment.js (hunk 23 above)
Comment (josemarinas) [reply packages/js-client/test-environment.js:3]: Yes, we added this at the begining so I dont remember exactly but I think we used some browser only sutuff from js and we needed to addd the shims fopr the tests to work on the nodejs env

File: packages/js-client/test/mocks/aragon-sdk-ipfs.ts (hunk 24 above)
Comment (josemarinas) [reply packages/js-client/test/mocks/aragon-sdk-ipfs.ts:11]: 👍

File: packages/js-client/test/mocks/graphql-request.ts (hunk 25 above)
Comment (josemarinas) [reply packages/js-client/test/mocks/graphql-request.ts:13]: 👍

File: packages/js-client/test/unit/context.test.ts (hunk 26 above)
Comment (josemarinas) [reply packages/js-client/test/unit/context.test.ts:90]: 👍

File: .github/workflows/js-client-update-docs.yml (hunk 7 above)
Comment (brickpop) [reply .github/workflows/js-client-update-docs.yml]: Maybe we could leave these lines but commented out on the template?

File: .github/workflows/js-client-update-docs.yml (hunk 8 above)
Comment (brickpop) [reply .github/workflows/js-client-update-docs.yml]: Commenting the irrelevant bits out?

File: packages/js-client/package.json (hunk 11 above)
Comment (brickpop) [reply packages/js-client/package.json]: That works too, but not "sdk"

File: packages/js-client/src/context.ts (hunk 13 above)
Comment (brickpop) [reply packages/js-client/src/context.ts]: But what are we importing?

File: packages/js-client/src/context.ts (hunk 14 above)
Comment (brickpop) [reply packages/js-client/src/context.ts]: My point is that it looks like we are defining the same type twice, in different ways.

See `packages/js-client/src/types.ts` below

File: packages/js-client/src/context.ts (hunk 15 above)
Comment (brickpop) [reply packages/js-client/src/context.ts]: Having the custom settings of the plugin on an object called `pluginContext` and the general settings from Aragon's `Context` called `aragonContext` or `globalContext`.

Or maybe a generic version of `Context<T>` in just one variable

File: packages/js-client/package.json [hunk 27]
```diff
@@ -0,0 +1,75 @@
...
+    "singleQuote": true,
+    "trailingComma": "es5"
+  },
+  "size-limit": [
+    {
+      "path": "dist/sdk-client-template.cjs.production.min.js",
+      "limit": "10 KB"
+    },
+    {
+      "path": "dist/sdk-client-template.esm.js",
```
Comment (brickpop) [inline packages/js-client/package.json]: The artifact names should also reflect the intended `js-client-template`

File: packages/js-client/src/client.ts [hunk 28]
```diff
@@ -0,0 +1,31 @@
...
+  public estimation: ISimpleStorageClientEstimation;
+  public encoding: ISimpleStorageClientEncoding;
+  public decoding: ISimpleStorageClientDecoding;
+
+  constructor(context: SimpleStorageContext) {
+    super(context);
+    this.methods = new SimpleStorageClientMethods(context);
+    this.estimation = new SimpleStoragClientEstimation(context);
+    this.encoding = new SimpleStorageClientEncoding(context);
+    this.decoding = new SimpleStorageClientDecoding(context);
```
Comment (brickpop) [inline packages/js-client/src/client.ts]:
```suggestion
    // Go to the implementation of each class to add your custom methods there
    this.methods = new SimpleStorageClientMethods(context);
    this.estimation = new SimpleStoragClientEstimation(context);
    this.encoding = new SimpleStorageClientEncoding(context);
    this.decoding = new SimpleStorageClientDecoding(context);
```

File: packages/js-client/src/types.ts (hunk 22 above)
Comment (brickpop) [reply packages/js-client/src/types.ts]: See `packages/js-client/src/context.ts`. There's the same local definition, but for state and overridden

File: packages/js-client/test/unit/context.test.ts (hunk 26 above)
Comment (josemarinas) [reply packages/js-client/test/unit/context.test.ts:90]: wait not necesarily because the supported network is something more for internal use, here we can receive a networkish, wich is the name, or the number of the network

File: packages/contracts-ethers/npm/index.ts [hunk 29]
```diff
@@ -1,5 +1,2 @@
 // TODO
```
Comment (brickpop) [inline packages/contracts-ethers/npm/index.ts]: Needed?

```suggestion
// TODO
```

File: packages/contracts-ethers/package.json [hunk 30]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@aragon/todo-plugin-ethers",
```
Comment (brickpop) [inline packages/contracts-ethers/package.json]:
```suggestion
  "name": "@aragon/plugin-template-ethers",
```

File: packages/contracts-ethers/package.json [hunk 31]
```diff
@@ -1,6 +1,6 @@
...
-  "version": "1.2.1",
+  "version": "1.0.0",
   "description": "The Aragon OSx contract definitions for ethers.js",
```
Comment (brickpop) [inline packages/contracts-ethers/package.json]:
```suggestion
  "description": "Plugin contract definitions for ethers.js",
```

File: packages/js-client/src/context.ts [hunk 32]
```diff
@@ -0,0 +1,77 @@
...
+    // parent contructor will call this.set(contextParams)
+    // so we don't need to call it again
+    super(contextParams);
+    if (aragonContext) {
+      // copy the context properties to this
+      Object.assign(this, aragonContext);
+    }
+    if (contextParams) {
+      // overide the context params with the ones passed to the constructor
+      this.set(contextParams ?? {});
```
Comment (brickpop) [inline packages/js-client/src/context.ts]: Not needed since we are within the "if" block

```suggestion
      this.set(contextParams);
```

File: packages/js-client/src/context.ts [hunk 33]
```diff
@@ -0,0 +1,77 @@
...
+  constructor(
+    contextParams?: Partial<SimpleStorageContextParams>,
+    aragonContext?: Context,
+  ) {
+    // parent contructor will call this.set(contextParams)
+    // so we don't need to call it again
+    super(contextParams);
+    if (aragonContext) {
+      // copy the context properties to this
+      Object.assign(this, aragonContext);
```
Comment (brickpop) [inline packages/js-client/src/context.ts:32]: Wouldn't this replace `this.state` and `this.overridden` completely and wipe the data set on the `super(contextParams)` line?

File: packages/js-client/src/internal/interfaces.ts [hunk 34]
```diff
@@ -0,0 +1,35 @@
...
+  ): AsyncGenerator<PrepareInstallationStepValue>;
+  storeNumber(number: bigint): AsyncGenerator<StoreNumberStepValue>;
+}
+export interface ISimpleStorageClientEstimation {
+  prepareInstallation(
+    params: PrepareInstallationParams,
+  ): Promise<GasFeeEstimation>;
+  storeNumber(
+    number: bigint,
+  ): Promise<GasFeeEstimation>;
```
Comment (brickpop) [inline packages/js-client/src/internal/interfaces.ts]: Don't we have a method to retrieve the stored value?

File: packages/js-client/src/types.ts [hunk 35]
```diff
@@ -0,0 +1,58 @@
+import {
+  ContextParams,
+  ContextState,
+  // OverridenState,
+  VersionTag,
+} from "@aragon/sdk-client-common";
+
+// TODO
+// import from @aragon/sdk-client-common
+type OverridenState = {
```
Comment (brickpop) [inline packages/js-client/src/types.ts]: Why don't we pull this from client common?

File: packages/js-client/test/integration/decoding.test.ts [hunk 36]
```diff
@@ -0,0 +1,21 @@
...
+  });
+
+  afterAll(async () => {
+    server.close();
+  });
+
+  it('should decode an action', async () => {
+    const ctx = new SimpleStorageContext();
+    const client = new SimpleStorageClient(ctx);
+    client.decoding.myAction();
```
Comment (brickpop) [inline packages/js-client/test/integration/decoding.test.ts]: Does this function still exist? Tests should be failing here

File: packages/js-client/test/integration/encoding.test.ts [hunk 37]
```diff
@@ -0,0 +1,21 @@
...
+  });
+
+  afterAll(async () => {
+    server.close();
+  });
+
+  it('should encode an action', async () => {
+    const ctx = new SimpleStorageContext();
+    const client = new SimpleStorageClient(ctx);
+    client.encoding.myAction();
```
Comment (brickpop) [inline packages/js-client/test/integration/encoding.test.ts]: Same as above

File: packages/js-client/test/unit/context.test.ts [hunk 38]
```diff
@@ -0,0 +1,291 @@
...
+    // reset contextParams
+    contextParams = {
+      network: 'mainnet',
+      signer: new Wallet(TEST_WALLET),
+      daoFactoryAddress: '0x1234',
+      web3Providers: web3endpoints.working,
+      gasFeeEstimationFactor: 0.1,
+      graphqlNodes: [],
+      ipfsNodes: [],
+      simpleStoragePluginAddress: 'notDefault',
```
Comment (brickpop) [inline packages/js-client/test/unit/context.test.ts]:
```suggestion
      // General parameters
      network: 'mainnet',
      signer: new Wallet(TEST_WALLET),
      daoFactoryAddress: '0x1234',
      web3Providers: web3endpoints.working,
      gasFeeEstimationFactor: 0.1,
      graphqlNodes: [],
      ipfsNodes: [],
      // Plugin specific parameters
      simpleStoragePluginAddress: 'notDefault',
```

File: packages/js-client/test/unit/context.test.ts (hunk 26 above)
Comment (brickpop) [reply packages/js-client/test/unit/context.test.ts:90]: Isn't this parameter what client common uses for applying the defaults? These defaults are directly related to our `SupportedNetworks`

File: packages/js-client/test/unit/context.test.ts [hunk 39]
```diff
@@ -0,0 +1,291 @@
...
+    );
+    context.graphql?.map((graphqlClient) =>
+      expect(graphqlClient).toBeInstanceOf(GraphQLClient)
+    );
+    expect(context.gasFeeEstimationFactor).toEqual(0.1);
+    expect(context.simpleStoragePluginAddress).toBe('notDefault');
+  });
+  it('Should create a context in goerli, update the network and update all the parameters automatically', () => {
+    const context = new SimpleStorageContext({
+      network: 'goerli',
```
Comment (brickpop) [inline packages/js-client/test/unit/context.test.ts:120]: Same

File: packages/js-client/test/unit/context.test.ts [hunk 40]
```diff
@@ -0,0 +1,291 @@
...
+    context.ipfs.map((ipfsClient) => {
+      expect(ipfsClient).toBeInstanceOf(IpfsClient);
+    });
+    context.graphql.map((graphqlClient) =>
+      expect(graphqlClient).toBeInstanceOf(GraphQLClient)
+    );
+
+    expect(context.simpleStoragePluginAddress).toBe("0x1234567890123456789012345678901234567890");
+    context.set({
+      network: 'matic',
```
Comment (brickpop) [inline packages/js-client/test/unit/context.test.ts:146]: +1

File: packages/js-client/test/unit/context.test.ts (hunk 26 above)
Comment (josemarinas) [reply packages/js-client/test/unit/context.test.ts:90]: Yes, but we support Inputing a `Networkish` then we do a `getNetwork` to transform it to our `SupportedNetworks`, else users would need to pass `homestrad` instead of `mainnet` and we wouldn't supoort passing the chain id

File: packages/contracts-ethers/npm/index.ts (hunk 29 above)
Comment (josemarinas) [reply packages/contracts-ethers/npm/index.ts]: not sure what you mean, I just removed the imports that were not resolving

File: packages/js-client/src/context.ts (hunk 33 above)
Comment (josemarinas) [reply packages/js-client/src/context.ts:32]: Yep, that why i call it again inside the set function, a bit complex but I wanted to give priority to the context params over the context and TS does not allow you to have a contructor without calling super

File: packages/js-client/src/internal/interfaces.ts (hunk 34 above)
Comment (josemarinas) [reply packages/js-client/src/internal/interfaces.ts]: Because I forgot

File: packages/js-client/src/types.ts (hunk 35 above)
Comment (josemarinas) [reply packages/js-client/src/types.ts]: Its not exported, thats why the todo is there, I wanted to finish the template and gather the missing pieces from the sdk-common-client to add them all in one commit

File: packages/js-client/test/integration/decoding.test.ts (hunk 36 above)
Comment (josemarinas) [reply packages/js-client/test/integration/decoding.test.ts]: nope, tests fixes are on the way

File: packages/contracts-ethers/npm/index.ts (hunk 29 above)
Comment (brickpop) [reply packages/contracts-ethers/npm/index.ts]: Do we need to have a `// TODO` for the template?

File: packages/contracts-ethers/package.json (hunk 30 above)
Comment (brickpop) [reply packages/contracts-ethers/package.json]: ping

File: packages/js-client/package.json (hunk 3 above)
Comment (brickpop) [reply packages/js-client/package.json:21]: Is this still happening now?

File: packages/js-client/package.json (hunk 5 above)
Comment (brickpop) [reply packages/js-client/package.json:24]: Is the name of the artifacts the same? I assume you ran the build command

File: packages/js-client/package.json (hunk 6 above)
Comment (brickpop) [reply packages/js-client/package.json:25]: @josemarinas

File: packages/js-client/src/context.ts (hunk 33 above)
Comment (brickpop) [reply packages/js-client/src/context.ts:32]: This might be confusing for people approaching the template. Wouldn't a `@ts-ignore` silence this constructor+super warning, so that we did things in the right order?

Or maybe running `super({})` with a comment explaining that we do this to silence the warning, but not shifting the attention away from the right way of using the context

File: packages/js-client/src/core.ts [hunk 41]
```diff
@@ -0,0 +1,13 @@
+import { ClientCore } from "@aragon/sdk-client-common";
+import { SimpleStorageContext } from "./context";
+
+export class SimpleStorageClientCore extends ClientCore {
```
Comment (brickpop) [inline packages/js-client/src/core.ts]: What's the goal of this class? Don't we have the client class itself and the context class?

Is there a way to have state/settings + logic in two components instead of 3?

File: packages/js-client/src/internal/interfaces.ts (hunk 34 above)
Comment (brickpop) [reply packages/js-client/src/internal/interfaces.ts]: Let's add it if possible, please

File: packages/js-client/src/types.ts (hunk 35 above)
Comment (brickpop) [reply packages/js-client/src/types.ts]: I would personally favor getting things cohesive here before relying on other tasks to improve the base layout, when the mental context has shifted to something different

File: packages/js-client/test/integration/decoding.test.ts (hunk 36 above)
Comment (brickpop) [reply packages/js-client/test/integration/decoding.test.ts]: It's a good practice to prevent PR's from getting too big, bit in this case we are shipping the initial commit. Personally I would prefer having a good initial version instead creating many more PR's to amend easy wins that are in our mind now

File: packages/js-client/test/unit/context.test.ts (hunk 26 above)
Comment (brickpop) [reply packages/js-client/test/unit/context.test.ts:90]: I see
feel free to resolve then

File: packages/js-client/src/core.ts (hunk 41 above)
Comment (josemarinas) [reply packages/js-client/src/core.ts]: Mmmm I'm afraid we can't, we have the context and the client, but then we have the core that is the one that has all the web3, graphql and ipfs modules and starts the clients, so all the modules must implement this, the easier solution i found is extending this class and exporting here the new parameters and the implementing this class in the methods, encodign, decoding and estimation modules so we have access to the parameters

File: packages/js-client/src/context.ts (hunk 33 above)
Comment (josemarinas) [reply packages/js-client/src/context.ts:32]: a ts-ignoer wont solve anything because then it will complaain because you wouldn have access to `this`, we can leeave a `super()`

File: packages/js-client/package.json (hunk 3 above)
Comment (josemarinas) [reply packages/js-client/package.json:21]: nope

File: packages/js-client/package.json (hunk 6 above)
Comment (josemarinas) [reply packages/js-client/package.json:25]: This is working but we are waaay over the size limit

imagen

File: packages/js-client/package.json (hunk 5 above)
Comment (josemarinas) [reply packages/js-client/package.json:24]: this is solved

File: packages/js-client/src/context.ts (hunk 33 above)
Comment (brickpop) [reply packages/js-client/src/context.ts:32]: Up to you :+1:

File: packages/js-client/src/core.ts (hunk 41 above)
Comment (brickpop) [reply packages/js-client/src/core.ts]: Why don't we export the ClientCore class so that plugins don't need to reinvent the wheel each time?
It's true that some might not need subgraph, but then they are free to just build their own client

File: packages/js-client/package.json (hunk 6 above)
Comment (brickpop) [reply packages/js-client/package.json:25]: Is there any way to easily reduce what we export? If not, we'll need a separate task to optimize this

File: packages/js-client/package.json [hunk 42]
```diff
@@ -0,0 +1,81 @@
...
+    "trailingComma": "es5"
+  },
+  "size-limit": [
+    {
+      "path": "dist/simple-storage-js-client.cjs.production.min.js",
+      "limit": "10 KB"
+    },
+    {
+      "path": "dist/simple-storage-js-client.esm.js",
+      "limit": "10 KB"
```
Comment (brickpop) [inline packages/js-client/package.json]: Couldn't we raise this to more reasonable values?

File: packages/js-client/src/context.ts (hunk 33 above)
Comment (brickpop) [reply packages/js-client/src/context.ts:32]: The other part the comment was about the shallow copy being done here

If the values of aragonContext change, this class will be affected as well which is a side effect. Isn't it possible to do a destructured copy similar to `Object.assign(this, {...aragonContext})`? Or even just getting the `aragonContext` values and doing a manual `super.set(...)`?

File: packages/js-client/src/internal/modules/methods.ts [hunk 43]
```diff
@@ -0,0 +1,74 @@
...
+import { SubgraphNumberListItem, SubgraphNumber } from "../types";
+import { toNumber, toNumberListItem } from "../utils";
+
+export class SimpleStorageClientMethods extends SimpleStorageClientCore
+  implements ISimpleStorageClientMethods {
+  // implementation of the methods in the interface
+  public async *prepareInstallation(
+    params: PrepareInstallationParams,
+  ): AsyncGenerator<PrepareInstallationStepValue> {
+    yield* prepareGenericInstallation(this.web3, {
```
Comment (brickpop) [inline packages/js-client/src/internal/modules/methods.ts:31]:
```suggestion
    // do any additionall custom operations here before you prepare your plugin

    // ...
    
    yield* prepareGenericInstallation(this.web3, {
```

File: packages/js-client/src/core.ts (hunk 41 above)
Comment (josemarinas) [reply packages/js-client/src/core.ts]: The client core class is exported, in fact SimpleStorageCore, extends it

It contains all the retry logic, and creates all internal clients

File: packages/js-client/package.json (hunk 6 above)
Comment (josemarinas) [reply packages/js-client/package.json:25]: We can increase it to 1MB maybe. The ethers, osx-ethers and  graphql are pretty big packages so maybe the problem comes from there. Not sure

File: packages/js-client/src/types.ts [hunk 44]
```diff
@@ -0,0 +1,50 @@
...
+  // extend the Context state with a new state for storing
+  // the new parameters
+  simpleStoragePluginAddress: string;
+  simpleStorageRepoAddress: string;
+};
+
+export type SimpleStorageOverriddenState = OverriddenState & {
+  simpleStoragePluginAddress: boolean;
+  simpleStorageRepoAddress: boolean;
+};
```
Comment (brickpop) [inline packages/js-client/src/types.ts]: Avoid human error by automating the `boolean` type remap

```suggestion
export type SimpleStorageOverriddenState = OverriddenState & {
  [key in keyof SimpleStorageContextState]: boolean;
};
```
